From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SR_team <sr_team@sr.team>
Date: Mon, 13 Jun 2022 16:38:16 +0300
Subject: [PATCH] 4341: Add support second wheel for horizontal scroll

Adding support horizontal scroll by second wheel, 4d wheel and touchpad
---
 .../docking/widgets/fieldpanel/FieldPanel.java | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/Ghidra/Framework/Docking/src/main/java/docking/widgets/fieldpanel/FieldPanel.java b/Ghidra/Framework/Docking/src/main/java/docking/widgets/fieldpanel/FieldPanel.java
index 5659bf675..194cf8e0d 100644
--- a/Ghidra/Framework/Docking/src/main/java/docking/widgets/fieldpanel/FieldPanel.java
+++ b/Ghidra/Framework/Docking/src/main/java/docking/widgets/fieldpanel/FieldPanel.java
@@ -48,6 +48,9 @@ public class FieldPanel extends JPanel
 		implements IndexedScrollable, LayoutModelListener, ChangeListener, Accessible {
 	public static final int MOUSEWHEEL_LINES_TO_SCROLL = 3;
 
+	public static final int MOUSEWHEEL_SCROLL_LEFT_BTN = 4;
+	public static final int MOUSEWHEEL_SCROLL_RIGHT_BTN = 5;
+
 	private LayoutModel model;
 
 	private boolean repaintPosted;
@@ -1328,6 +1331,21 @@ public class FieldPanel extends JPanel
 				l.buttonPressed(loc, field, ev);
 			}
 		});
+		if ((ev.getButton() == MOUSEWHEEL_SCROLL_LEFT_BTN || ev.getButton() == MOUSEWHEEL_SCROLL_RIGHT_BTN)
+				&& horizontalScrollingEnabled) {
+			int wheelRotation = (ev.getButton() == MOUSEWHEEL_SCROLL_LEFT_BTN ? -1 : 1);
+			Layout firstLayout = model.getLayout(BigInteger.ZERO);
+			int layoutScrollHt = firstLayout != null //
+					? firstLayout.getScrollableUnitIncrement(0, 1)
+					: 0;
+			int scrollAmount = wheelRotation * layoutScrollHt * MOUSEWHEEL_LINES_TO_SCROLL;
+			MouseWheelEvent wev = new MouseWheelEvent(ev.getComponent(),ev.getID(), ev.getWhen(),
+					InputEvent.SHIFT_DOWN_MASK, ev.getX(), ev.getY(), ev.getClickCount(), ev.isPopupTrigger(),
+					MouseWheelEvent.WHEEL_UNIT_SCROLL, scrollAmount, wheelRotation);
+			for (int i = 0; i < getMouseWheelListeners().length; i++) {
+				getMouseWheelListeners()[i].mouseWheelMoved(wev);
+			}
+		}
 	}
 
 	private JViewport getViewport() {
-- 
2.40.1

